stdlib/public/core/UnsafeRawBufferPointer.swift.gyb

```diff
@@ -316,6 +316,41 @@ extension Unsafe${Mutable}RawBufferPointer: ${Mutable}Collection {
     }
     return 0
   }
+
+  public func _customIndexOfEquatableElement(_ element: UInt8) -> Index?? {
+    guard !isEmpty else { return Optional(nil) }
+
+    let ptr = baseAddress._unsafelyUnwrappedUnchecked
+    var i = 0
+
+    let stride = MemoryLayout<UInt>.stride
+    let address = Int(bitPattern: ptr)
+
+    while (address &+ i) % stride != 0 && i < count {
+      guard self[i] != element else { return Optional(i) }
+      i &+= 1
+    }
+
+    var mask = UInt(0)
+    withUnsafeMutableBytes(of: &mask) {
+      $0.initializeMemory(as: UInt8.self, repeating: element)
+    }
+
+    while (i &+ stride) <= count {
+      let word: UInt = UnsafePointer<UInt>(
+          bitPattern: address &+ i
+        )._unsafelyUnwrappedUnchecked.pointee
```

Comment:
Would `ptr.load(fromByteOffset: i, as: UInt.self)` generate worse code?

Reply:
Or, if we _must_ resort to perhaps-undefined behaviour:
`let word = ptr.advanced(by: i).assumingMemoryBound(to: UInt.self).pointee`

Reply:
For the same short instruction count, but helping the compiler understand what's going on:
```swift
      let word = ptr.advanced(by: i).withMemoryRebound(
        to: UInt.self, capacity: 1, { $0.pointee }
      )
```
(80 columns is very narrow.)
Also, `UInt` isn't 64 bits on every platform. It should be `UInt64`.

Reply:
All three fixes you propose generate the same code of two move statements on ARM. I personally find variant `ptr.load(fromByteOffset: i, as: UInt.self)` the easiest to understand. I don't get why last variant you proposed is helping compiler understand what's going on and why should we care about this.

About `UInt64`. Why not `UInt128` then? I expect `UInt` be the fastest on all platforms (32 or 64). Or probably, the opposite - `UInt64` might be slower on 32-bit platform. I don't have 32 bit platform to benchmark this. How should we proceed with this? Should I make separate PR with `UInt64` variant and after benchmarking we could make our choice?

Reply:
Regarding `ptr.load(fromByteOffset: i, as: UInt.self)`, I happened to be looking at x86_64, and the `load` option generates two extra instructions when compared to the straight pointer dereference. 🤷🏽 The last variant I suggested is better than your original one because bouncing between `Int` and pointer representations confuses the compiler. Did you try using `loadUnaligned`? If that generates compact code then we can eliminate the first loop that brings the pointer into alignment.

Reply:
> Did you try using loadUnaligned? If that generates compact code then we can eliminate the first loop that brings the pointer into alignment.

Surprisingly, `loadUnaligned` is the same one move statement (checked on X86-64). But if buffer is shorter than word width, it will be segfault (will be?). So we have to branch here depending on buffer length. Does it worth it?

Reply:
It's certainly worth skipping when the buffer is too narrow: we clearly want to avoid the out-of-bounds memory access.

Reply:
Something to look for is that there may be a minimum number of strides for which it's worth paying the setup cost of doing a vectorized loop. I would hope that number to be 1, but I wouldn't be surprised by a different result.